src/intersections/sweepline.jl

```diff
@@ -0,0 +1,240 @@
+      if new_type == IntersectionType(0)
+        BinaryTrees.insert!(𝒬, new_geom)
+        haskey(𝒞, new_geom) ? push!(𝒞[new_geom], _key(below), _segdata(s)) : (𝒞[new_geom] = [_key(below), _segdata(s)])
+      end
+    end
+    if above !== nothing
+      new_geom, new_type = _newevent(s, Segment(_key(above)))
+      if new_type == IntersectionType(0)
+        BinaryTrees.insert!(𝒬, new_geom)
+        haskey(𝒞, new_geom) ? push!(𝒞[new_geom], _segdata(s), _key(above)) : (𝒞[new_geom] = [_segdata(s), _key(above)])
+      end
+    end
+  end
+end
+
+function _process_end_segments!(end_segments, 𝒬, 𝒯, 𝒞)
+  for s in end_segments
+    above, below = find_above_below(BinaryTrees.root(𝒯), BinaryTrees.search(𝒯, s))
+    BinaryTrees.delete!(𝒯, s)
+    s = Segment(s)
+    if above !== nothing && below !== nothing
+      new_geom, new_type = _newevent(Segment(_key(above)), Segment(_key(below)))
+      if new_type == IntersectionType(0)
+        BinaryTrees.insert!(𝒬, new_geom)
+        haskey(𝒞, new_geom) ? push!(𝒞[new_geom], _key(above), _key(below)) : (𝒞[new_geom] = [_key(above), _key(below)])
+      end
+    end
+  end
+end
+
+function _process_intersection_segments!(intersection_segments, 𝒬, 𝒯, 𝒞)
+  for s in intersection_segments
+    _, below = find_above_below(BinaryTrees.root(𝒯), BinaryTrees.search(𝒯, s))
+    if below !== nothing
+      # Swap positions of s and t in 𝒯
+      BinaryTrees.delete!(𝒯, s)
+      BinaryTrees.delete!(𝒯, _key(below))
+      BinaryTrees.insert!(𝒯, _key(below))
+      BinaryTrees.insert!(𝒯, s)
+
+      # Find segments r and u
+      _, r = find_above_below(BinaryTrees.root(𝒯), BinaryTrees.search(𝒯, _key(below)))
+      u, _ = find_above_below(BinaryTrees.root(𝒯), BinaryTrees.search(𝒯, s))
+
+      # Remove crossing points rs and tu from event queue
+      if r !== nothing
+        new_geom, new_type = _newevent(Segment(_key(r)), Segment(s))
+        if new_type == IntersectionType(0)
+          BinaryTrees.delete!(𝒬, new_geom)
+        end
+      end
+      if u !== nothing
+        new_geom, new_type = _newevent(Segment(_key(u)), Segment(_key(below)))
+        if new_type == IntersectionType(0)
+          BinaryTrees.delete!(𝒬, new_geom)
+        end
+      end
+
+      # Add crossing points rt and su to event queue
+      if r !== nothing
+        new_geom, new_type = _newevent(Segment(_key(r)), Segment(_key(below)))
+        if new_type == IntersectionType(0)
+          BinaryTrees.insert!(𝒬, new_geom)
+          haskey(𝒞, new_geom) ? push!(𝒞[new_geom], _key(r), _key(below)) : (𝒞[new_geom] = [_key(r), _key(below)])
+        end
+      end
+      if u !== nothing
+        new_geom, new_type = _newevent(Segment(_key(u)), Segment(s))
+        if new_type == IntersectionType(0)
+          BinaryTrees.insert!(𝒬, new_geom)
+          haskey(𝒞, new_geom) ? push!(𝒞[new_geom], _key(u), s) : (𝒞[new_geom] = [_key(u), s])
+        end
+      end
+    end
+  end
+end
+
+_segdata(seg::Segment) = seg.vertices.data
```

Comment:
We don't access fields directly. There is always a public api function `vertices(seg)` to get what we need. If there is no such function, then we probably need to add something before the PR.